syntax = "proto3";

package face_verification;

// Face verification service
service FaceVerification {
  // Enroll a face with single image
  rpc EnrollFace(EnrollRequest) returns (EnrollResponse);
  
  // Verify face with single image (1:1 or 1:N)
  rpc VerifyFace(VerifyRequest) returns (VerifyResponse);
  
  // Verify face with multiple frames (more robust)
  rpc VerifyFaceMultiFrame(VerifyMultiFrameRequest) returns (VerifyResponse);
  
  // Get user face profiles
  rpc GetUserProfiles(GetProfilesRequest) returns (GetProfilesResponse);
  
  // Update face profile
  rpc UpdateProfile(UpdateProfileRequest) returns (UpdateProfileResponse);
  
  // Delete face profile
  rpc DeleteProfile(DeleteProfileRequest) returns (DeleteProfileResponse);
}

// Messages for enrollment
message EnrollRequest {
  string user_id = 1;
  string company_id = 2;
  bytes image_data = 3;  // Image bytes (JPEG/PNG)
  optional string device_id = 4;
  bool make_primary = 5;
  map<string, string> metadata = 6;
}

message EnrollResponse {
  string status = 1;  // ok, duplicate, failed
  optional string profile_id = 2;
  optional string message = 3;
  repeated DuplicateProfile duplicate_profiles = 4;
  optional float quality_score = 5;
}

message DuplicateProfile {
  string user_id = 1;
  float similarity = 2;
}

// Messages for verification
message VerifyRequest {
  bytes image_data = 1;  // Image bytes
  string company_id = 2;
  optional string user_id = 3;  // For 1:1 verification
  optional string device_id = 4;
  string search_mode = 5;  // "1:1" or "1:N"
  int32 top_k = 6;
}

// Multi-frame verification request
message VerifyMultiFrameRequest {
  repeated bytes frames = 1;  // 3-5 image frames
  string company_id = 2;
  optional string user_id = 3;  // For 1:1 verification
  optional string device_id = 4;
  string search_mode = 5;  // "1:1" or "1:N"
  int32 top_k = 6;
}

message VerifyResponse {
  string status = 1;  // match, no_match, failed
  bool verified = 2;
  repeated Match matches = 3;
  optional Match best_match = 4;
  optional string message = 5;
  optional float liveness_score = 6;
}

message Match {
  string user_id = 1;
  string profile_id = 2;
  float similarity = 3;
  float confidence = 4;
  bool is_primary = 5;
}

// Messages for profile management
message GetProfilesRequest {
  string user_id = 1;
  string company_id = 2;
}

message GetProfilesResponse {
  repeated FaceProfile profiles = 1;
}

message FaceProfile {
  string profile_id = 1;
  string user_id = 2;
  string company_id = 3;
  string embedding_version = 4;
  bool is_primary = 5;
  string created_at = 6;
  string updated_at = 7;
  optional string deleted_at = 8;
  map<string, string> metadata = 9;
  optional float quality_score = 10;
}

message UpdateProfileRequest {
  string profile_id = 1;
  string company_id = 2;
  optional bytes image_data = 3;
  optional bool make_primary = 4;
  map<string, string> metadata = 5;
}

message UpdateProfileResponse {
  string status = 1;
  string message = 2;
}

message DeleteProfileRequest {
  string profile_id = 1;
  string company_id = 2;
  bool hard_delete = 3;
}

message DeleteProfileResponse {
  string status = 1;
  string message = 2;
}
